presto-parquet/src/main/java/io/prestosql/parquet/writer/repdef/DefLevelIterables.java

```diff
@@ -0,0 +1,306 @@
+                }
+
+                @Override
+                protected Optional<Integer> computeNext()
+                {
+                    if (iterator != null && iterator.hasNext()) {
+                        return iterator.next();
+                    }
+                    position++;
+                    if (position == columnarMap.getPositionCount()) {
+                        return endOfData();
+                    }
+                    if (columnarMap.isNull(position)) {
+                        return Optional.of(maxDefinitionLevel - 2);
+                    }
+                    else {
+                        int arrayLength = columnarMap.getEntryCount(position);
+                        if (arrayLength == 0) {
+                            return Optional.of(maxDefinitionLevel - 1);
+                        }
+                        iterator = new FixedValueIterator(arrayLength);
+                        return iterator.next();
+                    }
+                }
+            };
+        }
+    }
+
+    static class ColumnArrayDefLevelIterable
+            implements DefLevelIterable
+    {
+        private final ColumnarArray columnarArray;
+        private final int maxDefinitionLevel;
+
+        ColumnArrayDefLevelIterable(ColumnarArray columnarArray, int maxDefinitionLevel)
+        {
+            this.columnarArray = requireNonNull(columnarArray, "columnarArray is null");
+            this.maxDefinitionLevel = maxDefinitionLevel;
+        }
+
+        @Override
+        public DefLevelIterator getIterator()
+        {
+            return new DefLevelIterator()
+            {
+                private int position = -1;
+                private FixedValueIterator iterator;
+
+                @Override
+                boolean end()
+                {
+                    return iterator == null || !iterator.hasNext();
+                }
+
+                @Override
+                protected Optional<Integer> computeNext()
+                {
+                    if (iterator != null && iterator.hasNext()) {
+                        return iterator.next();
+                    }
+                    position++;
+                    if (position == columnarArray.getPositionCount()) {
+                        return endOfData();
+                    }
+                    if (columnarArray.isNull(position)) {
+                        return Optional.of(maxDefinitionLevel - 2);
+                    }
+                    else {
+                        int arrayLength = columnarArray.getLength(position);
+                        if (arrayLength == 0) {
+                            return Optional.of(maxDefinitionLevel - 1);
+                        }
+                        iterator = new FixedValueIterator(arrayLength);
+                        return iterator.next();
+                    }
+                }
+            };
+        }
+    }
+
+    static class FixedValueIterator
```

Comment:
This can be replaces with Collections `nCopies(length, Optional.empty()).iterator()`